<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>DeBruijn &mdash; DeBruijn 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Meteor" href="README.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            DeBruijn
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Summary:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Meteor</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Source code</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">DeBruijn</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">DeBruijn</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="debruijn">
<h1>DeBruijn<a class="headerlink" href="#debruijn" title="Permalink to this heading"></a></h1>
<section id="meteor">
<h2>Meteor<a class="headerlink" href="#meteor" title="Permalink to this heading"></a></h2>
<p><a class="reference external" href="http://bioconda.github.io/recipes/metaphlan/README.html"><img alt="install with bioconda" src="https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat" /></a>
<a class="reference external" href="https://forgemia.inra.fr/metagenopolis/meteor/-/releases"><img alt="Latest Release" src="https://forgemia.inra.fr/metagenopolis/meteor/-/badges/release.svg" /></a>
<a class="reference external" href="https://forgemia.inra.fr/metagenopolis/meteor/-/commits/dev"><img alt="pipeline status" src="https://forgemia.inra.fr/metagenopolis/meteor/badges/dev/pipeline.svg" /></a>
<a class="reference external" href="https://forgemia.inra.fr/metagenopolis/meteor/-/commits/dev"><img alt="coverage report" src="https://forgemia.inra.fr/metagenopolis/meteor/badges/dev/coverage.svg" /></a>
<a class="reference external" href="https://forgemia.inra.fr/metagenopolis/meteor/-/jobs/artifacts/dev/raw/pylint/pylint.log?job=pylint"><img alt="pylint" src="https://forgemia.inra.fr/metagenopolis/meteor/-/jobs/artifacts/dev/raw/pylint/pylint.svg?job=pylint" /></a></p>
<section id="introduction">
<h3>Introduction<a class="headerlink" href="#introduction" title="Permalink to this heading"></a></h3>
<p>Meteor is a plateform for quantitative metagenomics profiling of complex ecosystems.
Meteor relies on genes catalogue to perform specie level taxonomic assignments and functional analysis.</p>
<p>Check the <a class="reference external" href="https://forgemia.inra.fr/metagenopolis/meteor/-/wikis/home">wiki</a> for more information.
If you use meteor , please cite:</p>
</section>
<section id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this heading"></a></h3>
<p>Meteor requires:</p>
<ul class="simple">
<li><p>python&gt;3.11.0</p></li>
<li><p><a class="reference external" href="https://github.com/BenLangmead/bowtie2">bowtie2</a></p></li>
<li><p><a class="reference external" href="https://pysam.readthedocs.io/en/latest/installation.html">pysam</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html">pandas</a></p></li>
</ul>
</section>
<section id="installation">
<h3>Installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h3>
<p>Meteor is available with bioconda which includes all its dependencies:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">meteor</span>
</pre></div>
</div>
<p>And with pip with a recent Python &gt;= 3.11:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">meteor</span>
</pre></div>
</div>
<p>You can test the installation of meteor with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">test</span>
</pre></div>
</div>
</section>
<section id="getting-started">
<h3>Getting started<a class="headerlink" href="#getting-started" title="Permalink to this heading"></a></h3>
<p>A basic usage of meteor will require to:</p>
<ol class="arabic simple">
<li><p><strong>Download or build a reference catalogue</strong></p></li>
<li><p><strong>Import the raw fastq files</strong></p></li>
<li><p><strong>Map reads against the reference catalogue</strong></p></li>
<li><p><strong>Profile taxonomical or functionnal abundances</strong></p></li>
</ol>
<section id="download-or-build-a-custom-reference">
<h4>1. Download or build a custom reference<a class="headerlink" href="#download-or-build-a-custom-reference" title="Permalink to this heading"></a></h4>
<hr class="docutils" />
<p>Meteor requires to download locally a microbial gene catalogue. Several catalogues are currently available:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head text-left"><p>Microbial gene catalogue</p></th>
<th class="head text-center"><p>&lt;name&gt;</p></th>
<th class="head text-center"><p>Genes count (M)</p></th>
<th class="head text-center"><p>Metagenomic Species Pan-genomes (MSPs)</p></th>
<th class="head text-center"><p>Size (GB)</p></th>
<th class="head text-center"><p>Taxonomy catalogue size (MB)</p></th>
<th class="head text-center"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-left"><p><em>Gallus gallus domesticus</em></p></td>
<td class="text-center"><p>chicken_caecal</p></td>
<td class="text-center"><p>13.6</p></td>
<td class="text-center"><p>2420</p></td>
<td class="text-center"><p>15.5</p></td>
<td class="text-center"><p>628</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/FHPJH5">link</a></p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><em>Homo sapiens oral</em></p></td>
<td class="text-center"><p>human_oral</p></td>
<td class="text-center"><p>8.4</p></td>
<td class="text-center"><p>853</p></td>
<td class="text-center"><p>10.1</p></td>
<td class="text-center"><p>179</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/WQ4UTV">link</a></p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><em>Homo sapiens gut</em></p></td>
<td class="text-center"><p>human_gut</p></td>
<td class="text-center"><p>10.4</p></td>
<td class="text-center"><p>1990</p></td>
<td class="text-center"><p>9.3</p></td>
<td class="text-center"><p>391</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/FLANUP">link</a></p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><em>Mus musculus</em></p></td>
<td class="text-center"><p>mouse_gut</p></td>
<td class="text-center"><p>5.0</p></td>
<td class="text-center"><p>1252</p></td>
<td class="text-center"><p>5.5</p></td>
<td class="text-center"><p>347</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/L11MXM">link</a></p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><em>Oryctolagus cuniculus</em></p></td>
<td class="text-center"><p>rabbit_gut</p></td>
<td class="text-center"><p>5.7</p></td>
<td class="text-center"><p>1053</p></td>
<td class="text-center"><p>5.9</p></td>
<td class="text-center"><p>199</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/5EJKAS">link</a></p></td>
</tr>
<tr class="row-odd"><td class="text-left"><p><em>Rattus norvegicus</em></p></td>
<td class="text-center"><p>rat_gut</p></td>
<td class="text-center"><p>5.9</p></td>
<td class="text-center"><p>1627</p></td>
<td class="text-center"><p>5.1</p></td>
<td class="text-center"><p>348</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.57745/GVL2EE">link</a></p></td>
</tr>
<tr class="row-even"><td class="text-left"><p><em>Sus domesticus</em></p></td>
<td class="text-center"><p>pig_gut</p></td>
<td class="text-center"><p>9.3</p></td>
<td class="text-center"><p>1523</p></td>
<td class="text-center"><p>8.4</p></td>
<td class="text-center"><p>378</p></td>
<td class="text-center"><p><a class="reference external" href="https://entrepot.recherche.data.gouv.fr/dataset.xhtml?persistentId=doi:10.15454/OPAULL">link</a></p></td>
</tr>
</tbody>
</table>
<p>These references can be downloaded with following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">download</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">name</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">c</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">refdir</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>We also created smaller catalogue designed exclusively for taxonomical profiling. They are available with the tag (-t) :</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">download</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">name</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">c</span> <span class="o">-</span><span class="n">fast</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">refdir</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>Users can also import custom gene catalogue with the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">build</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">fastafile</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">n</span> <span class="o">&lt;</span><span class="n">name</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">refdir</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">t</span> <span class="o">&lt;</span><span class="n">threads</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="import-fastq">
<h4>2. Import fastq<a class="headerlink" href="#import-fastq" title="Permalink to this heading"></a></h4>
<hr class="docutils" />
<p>Meteor requires a first of fastq indexing:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">fastqdir</span><span class="o">&gt;</span>  <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">paired</span> <span class="n">reads</span><span class="p">]</span> <span class="o">-</span><span class="n">n</span> <span class="o">&lt;</span><span class="n">projectname</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">outputdir</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>When multiple sequencing are available for a library, the option -m allows to group these samples.
Example:</p>
<p>Illumina_lib1-<strong>SAMPLE_01</strong>.fastq <br />
Illumina_lib1-<strong>SAMPLE_02</strong>.fastq <br />
Illumina_lib2-<strong>SAMPLE_01</strong>.fastq <br />
Illumina_lib2-<strong>SAMPLE_02</strong>.fastq <br /></p>
<p>In this case, the following command will group these samples the same library:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">i</span> <span class="o">./</span>  <span class="o">-</span><span class="n">m</span> <span class="n">SAMPLE_</span>\\<span class="n">d</span><span class="o">+</span> <span class="o">-</span><span class="n">n</span> <span class="n">projectname</span> <span class="o">-</span><span class="n">o</span> <span class="n">outputdir</span>
</pre></div>
</div>
</section>
<section id="mapping">
<h4>3. Mapping<a class="headerlink" href="#mapping" title="Permalink to this heading"></a></h4>
<hr class="docutils" />
<p>The mapping is performed with the raw fastq file against a catalogue reference with the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">mapping</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">fastqdir</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">r</span> <span class="o">&lt;</span><span class="n">refdir</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">mappingdir</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="abundance-profiling">
<h4>4. Abundance profiling<a class="headerlink" href="#abundance-profiling" title="Permalink to this heading"></a></h4>
<hr class="docutils" />
<section id="taxonomical-analysis">
<h5><strong>Taxonomical analysis</strong><a class="headerlink" href="#taxonomical-analysis" title="Permalink to this heading"></a></h5>
<p>The taxonomical profiling can be performed at several level of accuracy (gene, MSP, SuperKingdom, Phylum, Class, Order, Family, Genus, Specie). MSPs were calculated with <a class="reference external" href="https://academic.oup.com/bioinformatics/article/35/9/1544/5106712">MSPminer</a>.
Their abundances can be profiled with the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">profiler</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">mappingdir</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">l</span> <span class="o">&lt;</span><span class="n">accuracy</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">countingdir</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
<section id="functionnal-analysis">
<h5><strong>Functionnal analysis</strong><a class="headerlink" href="#functionnal-analysis" title="Permalink to this heading"></a></h5>
<p>Meteor provides an ARDs annotation based on <a class="reference external" href="https://www.nature.com/articles/s41564-018-0292-6">MUSTARD</a> and a pathway annotation based on <a class="reference external" href="https://academic.oup.com/nar/article/36/suppl_1/D480/2507484">KEGG r107</a>. Their abundances can be profiled with the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">meteor</span> <span class="n">profiler</span> <span class="o">-</span><span class="n">i</span> <span class="o">&lt;</span><span class="n">mappingdir</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">a</span> <span class="o">&lt;</span><span class="n">annotation</span><span class="o">&gt;</span> <span class="o">-</span><span class="n">o</span> <span class="o">&lt;</span><span class="n">countingdir</span><span class="o">&gt;</span>
</pre></div>
</div>
</section>
</section>
</section>
<section id="the-meteor-team">
<h3>The METEOR team<a class="headerlink" href="#the-meteor-team" title="Permalink to this heading"></a></h3>
<p>The main contributors to METEOR:</p>
<ul class="simple">
<li><p>Franck Gauthier</p></li>
<li><p>Amine Ghozlane</p></li>
<li><p>Florian Plaza Oñate</p></li>
<li><p>Nicolas Pons</p></li>
<li><p>Florence Thirion</p></li>
</ul>
</section>
<section id="acknowledgements">
<h3>Acknowledgements<a class="headerlink" href="#acknowledgements" title="Permalink to this heading"></a></h3>
<p>Special thanks to the following people:</p>
<ul class="simple">
<li><p>Mathieu Almeida</p></li>
<li><p>Emmanuelle Le Chatelier</p></li>
</ul>
</section>
</section>
<div class="toctree-wrapper compound">
<p class="caption" role="heading"><span class="caption-text">Summary:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">Meteor</a><ul>
<li class="toctree-l2"><a class="reference internal" href="README.html#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#dependencies">Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#getting-started">Getting started</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#the-meteor-team">The METEOR team</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">Source code</a><ul>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.Color"><code class="docutils literal notranslate"><span class="pre">Color</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.get_arguments"><code class="docutils literal notranslate"><span class="pre">get_arguments()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.get_logging"><code class="docutils literal notranslate"><span class="pre">get_logging()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.isdir"><code class="docutils literal notranslate"><span class="pre">isdir()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.isfile"><code class="docutils literal notranslate"><span class="pre">isfile()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="modules.html#meteor.meteor.main"><code class="docutils literal notranslate"><span class="pre">main()</span></code></a></li>
</ul>
</li>
</ul>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="README.html" class="btn btn-neutral float-right" title="Meteor" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Institut Pasteur.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>